### ISSUE TITLE
Abstract Class / Polymorphism Support

### ISSUE THREAD

**Issue by Kirow** [realm/realm-cocoa#1109]
Does it possible to create smth like this:
![Abstract Data Model](http://i.stack.imgur.com/4J8Ed.png)

Blue - abstract classes. Such model will help me to incapsulate some logic and prevent code duplication.

In this example RLMItem can be associated with RLMCategory or/and RLMFolder. It would be very useful if i will be able to get all RLMFolders and RLMItems using [RLMCatItem allObjects]. 
Seems that it is not possible.

I've tried to make smth similar with subclasses, but as result - I have additional useless classes in schema.

This works well with CoreData, but can I expect smth like this in Realm?

**_In other words:**_
1) Abstract classes will not be displayed in Realm Browser
2) Ability to query abstract classes (this will include all subclass objects)

**Comment by puttin** [comment]
https://github.com/realm/realm-java/issues/761 is related issue

**Comment by nekonari** [comment]
@jpsim So what kind of inheritance/polymorphism *is* available in Realm at this time? From what I can see, with no polymorphism in querying, nor in relationships, there's basically _no inheritance_ supported.

**Comment by jpsim** [comment]
> So what kind of inheritance/polymorphism is available in Realm at this time? From what I can see, with no polymorphism in querying, nor in relationships, there's basically no inheritance supported.

Sorry for the late reply. Inheritance in Realm at the moment gets you:

* Class methods, instance methods and properties on parent classes are inherited in their child classes.
* Methods and functions that take parent classes as arguments can operate on subclasses.

It does not get you:

* Casting between polymorphic classes (subclass->subclass, subclass->parent, parent->subclass, etc.).
* Querying on multiple classes simultaneously.
* Multi-class containers (`RLMArray`/`List` and `RLMResults`/`Results`).

We're 100% behind adding this functionality in Realm, but as you can tell from the labels on this GH issue, it is neither a high priority for us at the moment, or easy to do. Some underlying architecture work is needed to move forward with these additional inheritance-related features.

In the meantime, you can work around these inheritance limitations in a number of ways:

## 1. Running queries on all related types and mapping back to arrays

```swift
class A: Object {
  dynamic var intProp = 0
}
class B: Object {}
class C: Object {}
extension Realm {
  func filter<ParentType: Object>(parentType parentType: ParentType.Type, subclasses: [ParentType.Type], predicate: NSPredicate) -> [ParentType] {
    return ([parentType] + subclasses).flatMap { classType in
      return self.objects(classType).filter(predicate).map { $0 }
    }
}

// Usage

let realm = try! Realm()
let allAClassesGreaterThanZero = realm.filter(A.self, [B.self, C.self], NSPredicate(format: "intProp > 0")) // => [A]
```

## 2. Using an option type for polymorphic relationships

```swift
class A: Object {
  dynamic var intProp = 0
}
class B: Object {}
class C: Object {}
class AClasses: Object {
  dynamic var a: A? = nil
  dynamic var b: B? = nil
  dynamic var c: C? = nil
}
class D: Object {
  dynamic var polymorphicA: AClasses? = nil
}
// D's polymorphicA value can hold a wrapped A, B or C object
```

## 3. Initializing objects with their polymorphic counterparts

Instead of casting, you can copy the underlying values from one object to another if they share those properties:

```swift
class A: Object {
  dynamic var intProp = 0
}
class B: Object {}

// Usage

let a = A(value: [42])
let b = B(value: a)
```

**Comment by mrackwitz** [comment]
## 4. Alternative: Using Composition instead of Inheritance

Instead of using inheritance, you can avoid it and it's current limitations with Realm  in some cases at all by composing your classes via linked objects.

```swift
class Animal: Object {
  dynamic var age = 0
}
class Duck : Object {
  dynamic var animal: Animal? = nil
  dynamic var name = ""
}
class Frog : Object {
  dynamic var animal: Animal? = nil
  dynamic var dateProp = NSDate()
}

// Usage
let duck = Duck(value [ animal: [ "age": 3 ], "name": "Gustav" ])
```

If you want to share behavior between multiple classes, you can e.g. facilitate Swift's default implementations of protocols:

```swift
protocol DuckType {
  dynamic var animal: Animal? { get }

  func quak() -> ()
}

extension DuckType {
  func quak() {
    for _ in 1...(animal?.age ?? 1) {
      print("quak")
    }
  }
}

extension Duck: DuckType {}
extension Frog: DuckType {}

// both can quak now
```

**Comment by wanbok** [comment]
@mrackwitz Let me ask you about `primaryKey`. If `Animal` has `id` as `primaryKey`, what property is the best to `Duck` and `Frog` for `primaryKey`?
`animal`, `animal`'s `id` or new `id`?

**Comment by mrackwitz** [comment]
@wanbok: You would need to manually take over the value of the `id` property on `Animal` into a property you defined in each of your classes, here `Duck` and `Frog`. Only 'string' and 'int' properties can be designated the primary key.

**Comment by JadenGeller** [comment]
## 5. Using a type-erased wrapper for polymorphic relationships

Using a type-erased wrapper is a scalable workaround for the current lack of support for polymorphic relationships. It even allows you to maintain your inheritance hierarchy.
```swift
class PaymentMethod: Object { } // Abstract class, though not necessary

class CreditCardPaymentMethod: PaymentMethod {
    dynamic var cardNumber: String = ""
    dynamic var csv: String = ""

    override static func primaryKey() -> String? {
        return "cardNumber"
    }
}

class PaypalPaymentMethod: PaymentMethod {
    dynamic var username: String = ""
    dynamic var password: String = ""

    override static func primaryKey() -> String? {
        return "username"
    }
}

// Define as many subclasses as you'd like—even sub-sub-sub classes!
```

If you want to store an instance of *any* subclass of `PaymentMethod`, define a type-erased wrapper that stores the type's name and the primary key.
```swift
class AnyPaymentMethod: Object {
    dynamic var typeName: String = ""
    dynamic var primaryKey: String = ""

    // A list of all subclasses that this wrapper can store
    static let supportedClasses: [PaymentMethod.Type] = [
        CreditCardPaymentMethod.self,
        PaypalPaymentMethod.self
    ]

    // Construct the type-erased payment method from any supported subclass
    convenience init(_ paymentMethod: PaymentMethod) {
        self.init()
        typeName = String(paymentMethod.dynamicType)
        guard let primaryKeyName = paymentMethod.dynamicType.primaryKey() else {
            fatalError("`\(typeName)` does not define a primary key")
        }
        guard let primaryKeyValue = paymentMethod.valueForKey(primaryKeyName) as? String else {
            fatalError("`\(typeName)`'s primary key `\(primaryKeyName)` is not a `String`")
        }
        primaryKey = primaryKeyValue
    }

    // Dictionary to lookup subclass type from its name
    static let methodLookup: [String : PaymentMethod.Type] = {
        var dict: [String : PaymentMethod.Type] = [:]
        for method in supportedClasses {
            dict[String(method)] = method
        }
        return dict
    }()

    // Use to access the *actual* PaymentMethod value, using `as` to upcast
    var value: PaymentMethod {
        guard let type = AnyPaymentMethod.methodLookup[typeName] else {
            fatalError("Unknown payment method `\(typeName)`")
        }
        guard let value = try! Realm().objectForPrimaryKey(type, key: primaryKey) else {
            fatalError("`\(typeName)` with primary key `\(primaryKey)` does not exist")
        }
        return value
    }
}
```

Now, we can create a type that stores an `AnyPaymentMethod`!
```swift
class Purchase: Object {
    dynamic var product: String = ""
    dynamic var price: Int = 0
    dynamic var paymentMethod: AnyPaymentMethod?
}
```

Let's check out how this is used in an example.
```swift
let realm = try! Realm()
try! realm.write {
    // Payment methods
    let creditCard = CreditCardPaymentMethod()
    creditCard.cardNumber = "314159265358979"
    creditCard.csv = "001"
    realm.add(creditCard)
    
    let paypal = PaypalPaymentMethod()
    paypal.username = "ilovepersistance"
    paypal.password = "swifty"
    realm.add(paypal)
    
    // Purchases
    let carPurchase = Purchase()
    carPurchase.product = "car"
    carPurchase.price = 35000
    carPurchase.paymentMethod = AnyPaymentMethod(creditCard)
    realm.add(carPurchase)
    
    let gamePurchase = Purchase()
    gamePurchase.product = "game"
    gamePurchase.price = 20
    gamePurchase.paymentMethod = AnyPaymentMethod(paypal)
    realm.add(gamePurchase)
}

// Later, we want to check the payment method, use the `value` property on `AnyPaymentMethod`
for purchase in realm.objects(Purchase) {
    if let creditCard = purchase.paymentMethod?.value as? CreditCardPaymentMethod {
        print("We bought a \(purchase.product) from credit card #\(creditCard.cardNumber)")
    } else if let paypal = purchase.paymentMethod?.value as? PaypalPaymentMethod {
        print("We bought a \(purchase.product) from paypal username @\(paypal.username)")
    } else {
        fatalError("Unknown payment method")
    }
}
```

Awesome! This workaround is well-suited for cases where there are many subclasses since, unlike the option type approach, it doesn't require a single property for each supported subclass. This means that you can add new subclasses without performing migrations!

### Caveats
* Inverse relationships on `PaymentMethod`s will not work properly. As a workaround, place the inverse relationship on `AnyPaymentMethod.
* `PaymentMethod`s will not be recursively added to Realm when an object containing an `AnyPaymentMethod` is added to Realm. Make sure you add each `PaymentMethod` individually.
* All subclasses supported by `AnyPaymentMethod` must use the *same* type of primary key (though it does not need to be `String`).

**Comment by paulfreeman** [comment]
Option 5 seemed the best workaround for me, but it was a shame to lose handy tricks like filtering on a list of items stored on an Object.  The example given is of a single AnyPaymentMethod reference, to me the need for polymorphism is more obvious with collections of varying types. 

Another benefit is so that in future if polymorphism is supported, it should be easier to refactor the application code to use standard filtering on a List, as the usage is localised to the call site rather than being expanded out into accessors elsewhere that unwrap the type erased Any type. 

I adapted the method from an earlier example so that I can make a call such as 

`
let myListOfPayments = someObject.payments 
let results:[CreditCardPaymentMethod] = realm!.filter(collection: myListOfPayments, type:CreditCardPaymentMethod.self, condition:NSPredicate(format:"owner == 'Jaden Geller'"))`
 
extension Realm  {
     /**
     *   filter type erased List
     */
    func filter<T>(collection:List<AnyPaymentMethod>, type:Object.Type, condition: NSPredicate  ) -> [ T ] {
        var results = [T]()
        let allfiltered =  objects(type).filter(condition)
        for component in collection {
            let val = component.value
            if let morphedComponent = val as? T {
                if allfiltered.contains(morphedComponent as! Object) {
                    results.append(morphedComponent)
                }
            }
        }
        return results
    }  
}

**Comment by StackHelp** [comment]
I have this kind of classes right now but no modify optional fieds removes it's data. How to retain that.

```
class MediaFile: Object {
    @objc dynamic var _id = ""
    @objc dynamic var userId = ""
    @objc dynamic var uploadId = ""
    @objc dynamic var mediaPath = ""
    @objc dynamic var groupId = ""
    @objc dynamic var resourceId = ""
    @objc dynamic var fileName = ""
    @objc dynamic var fileSize: Int64 = 0
    @objc dynamic var offset: Float = 0
    @objc dynamic var uploadedByte: Int64 = 0
    @objc dynamic var totalByte: Int64 = 0
    @objc dynamic var createdDate: Date?
    @objc dynamic var thumbnail: Data?
    @objc dynamic var status = ""
    @objc dynamic var addMedia: MediaAdd? = nil
    
    override static func primaryKey() -> String? {
        return "fileName"
    }
}

class MediaAdd: Object {
    @objc dynamic var title = ""
    @objc dynamic var desc = ""
    @objc dynamic var mediaSource = ""
    @objc dynamic var storageStatus = ""
    @objc dynamic var tags = ""
    @objc dynamic var category = 0
}
```

Here is what I add first time

```
MediaFile {
		_id = ;
		userId = 85;
		uploadId = E9CA5700-2974-4B0F-985E-426A48F6EE80;
		mediaPath = file:///var/mobile/Containers/Data/Application/9E1926C5-7D01-4813-9401-13D9C84B2742/Documents/img_637215759322598144.jpeg;
		groupId = CZ1dchWXat7k5RMA;
		resourceId = ;
		fileName = img_637215759322598144.jpeg;
		fileSize = 191607;
		offset = 100;
		uploadedByte = 191607;
		totalByte = 191607;
		createdDate = 2020-04-04 05:45:35 +0000;
		thumbnail = <length = 24, bytes = 0xffd8ffe000104a46494600010100000100010000ffdb0043 — 191607 total bytes>;
		status = completed;
		addMedia = MediaAdd {
			title = teat;
			desc = ;
			mediaSource = Bernama Images;
			storageStatus = FINAL;
			tags = ;
			category = 25;
		};
```

And on the update here is my code

                    ```
let mediaFile = MediaFile()
                    mediaFile.fileName = file["filename"].stringValue
                    mediaFile.userId = file["user_id"].stringValue
                    mediaFile._id = file["_id"].stringValue
                    mediaFile.mediaPath = file["mediaPath"].stringValue
                    mediaFile.fileSize = file["file_size"].int64Value
                    mediaFile.groupId = file["group_id"].stringValue
                    mediaFile.resourceId = file["resource_id"].stringValue
                    mediaFile.offset = file["offset"].floatValue
                    mediaFile.createdDate = file["createdDate"].stringValue.toDate()
                    mediaFile.status = file["status"].stringValue
                    
                    try! realm.write {
                        realm.add(mediaFile, update: .modified)
                    }
```

And Response after update

```
MediaFile {
	_id = 5e881f00138c8801cd3c288e;
	userId = 85;
	uploadId = ;
	mediaPath = file:///var/mobile/Containers/Data/Application/9E1926C5-7D01-4813-9401-13D9C84B2742/Documents/img_637215759322598144.jpeg;
	groupId = CZ1dchWXat7k5RMA;
	resourceId = 3c1afe93-ac87-4a91-9c23-80d9bcf6d703;
	fileName = img_637215759322598144.jpeg;
	fileSize = 191607;
	offset = 191607;
	uploadedByte = 0;
	totalByte = 0;
	createdDate = 2020-04-04 08:15:36 +0000;
	thumbnail = <(null) — 0 total bytes>;
	status = completed;
	addMedia = (null);
}
```

As you can see `addMedia` returns `nil` after an update as well as you can check `thumbnail` returns `nil` as well. How to retain both on update?

**Comment by toph-allen** [comment]
Is there any update on the status of this issue, or current recommendations for best practices? The Realm Swift documentation's [section on model inheritance](https://realm.io/docs/swift/latest/#model-inheritance) says that this feature is "on the roadmap" and links to this issue.

**Comment by eanagel** [comment]
6 years is a long time to be "on the roadmap", maybe it's time to get this one done?

**Comment by bdkjones** [comment]
7 years and counting.

**Comment by bdkjones** [comment]
@JadenGeller since it doesn't seem like Realm is going to improve anything here, it would be nice if you updated your type-erased wrapper example. It's 7 years out of date and throws all kinds of errors now.

Additionally, it doesn't look like it'll work with the new `@Persisted(primaryKey: true)` property wrapper. I'm doing this:

```
@Persisted(primaryKey: true) var uuid: UUID = UUID()
```

And that's crashing in `init()` in the guard statement that ensures we have a primaryKeyName.

**Comment by dianaafanador3** [comment]
Hi @bdkjones we still have `Polymorphic inheritance` on our roadmap, have in mind this is a complex and long project which takes development both on Core and SDK side.
As a workaround, I'm updating the type-erasure example posted on this issue for modern schema declaration.

``` swift
class PaymentMethod: Object {
    @Persisted var owner: String
}

class CreditCardPaymentMethod: PaymentMethod {
    @Persisted(primaryKey: true) var id: String = ObjectId.generate().stringValue
    @Persisted var cardNumber: String = ""
    @Persisted var csv: String = ""
}

class PaypalPaymentMethod: PaymentMethod {
    @Persisted(primaryKey: true) var id: String = ObjectId.generate().stringValue
    @Persisted var username: String = ""
    @Persisted var password: String = ""
}

class Purchase: Object {
    @Persisted(primaryKey: true) var id: ObjectId = ObjectId.generate()
    @Persisted var product: String = ""
    @Persisted var price: Int = 0
    @Persisted var paymentMethod: AnyPaymentMethod?
}

class AnyPaymentMethod: Object  {
    @Persisted(primaryKey: true) var primaryKey: String = ""
    @Persisted var typeName: String = ""

    // A list of all subclasses that this wrapper can store
    static let supportedClasses: [PaymentMethod.Type] = [
        CreditCardPaymentMethod.self,
        PaypalPaymentMethod.self
    ]

    // Construct the type-erased payment method from any supported subclass
    convenience init(_ paymentMethod: PaymentMethod) {
        self.init()
        typeName = String(describing: type(of: paymentMethod))
        guard let primaryKeyName = type(of: paymentMethod).sharedSchema()?.primaryKeyProperty?.name else {
            fatalError("`\(typeName)` does not define a primary key")
        }
        guard let primaryKeyValue = paymentMethod.value(forKey: primaryKeyName) as? String else {
            fatalError("`\(typeName)`'s primary key `\(primaryKeyName)` is not a `String`")
        }
        primaryKey = primaryKeyValue
    }

    // Dictionary to lookup subclass type from its name
    static let methodLookup: [String : PaymentMethod.Type] = {
        var dict: [String : PaymentMethod.Type] = [:]
        for method in supportedClasses {
            dict[String(describing: method)] = method
        }
        return dict
    }()

    // Use to access the *actual* PaymentMethod value, using `as` to upcast
    var value: PaymentMethod {
        guard let type = AnyPaymentMethod.methodLookup[typeName] else {
            fatalError("Unknown payment method `\(typeName)`")
        }
        guard let value = try! Realm().object(ofType: type, forPrimaryKey: primaryKey) else {
            fatalError("`\(typeName)` with primary key `\(primaryKey)` does not exist")
        }
        return value
    }
}

```
and can be used the same as proposed.

We will post on this issue any update regarding `Polymorphism` feature.

**Comment by bdkjones** [comment]
@dianaafanador3 Thanks for updating the workaround example. I DO understand that this is a large feature, but 8 years is a very long time and I don't think *realm* realizes just how long issues like this have been open.

Plus, polymorphism isn't some newfangled, fringe idea. It has existed for the better part of a century precisely because it's incredibly useful. And sure, the magic vomit in the example above is an ugly workaround, but when folks see issues like this go unaddressed for a decade, the thing they should be asking is: "Do I really want to tie myself to Realm?"

**Comment by ianpward** [comment]
Hi Product for Realm here - we realize how long this issue has been open, although perhaps that is a testament to just how long Realm has been a product and how many evolutions we've needed to go through over the years to get to this age. I will say that we do plan to do this because it would give us an incredibly unique position in the market, which, I would observe, there are no other cross-platform mobile databases that have Polymorphism/ Inheritance along with a backend database syncing solution, I believe this is evidence to how technically complex this feature is. We need to get this feature right across all of our SDKs + Sync + MongoDB.

**Comment by bdkjones** [comment]
@ianpward Thanks for the comment. I think it would be very helpful/reassuring if Realm had an official roadmap and *timetable* for major features like this. Eight years of "it's on the to-do list" comments really make me nervous about tying apps to Realm.

**Comment by marchy** [comment]
@bdkjones CC: @ianpward 
Agreed I think Realm could really benefit from a public roadmap.

[EF Core's roadmap](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-8.0/plan) is of particular influence. They've done a fantastic job of setting out the goals for the year ahead, putting out releases across the whole year, and calling out the things that won't make the cut for that year and re-prioritizing for the next major cycle.

I think Realm could benefit a lot from this strategy, and the entire feedback loop and trust-building that it creates.

**Comment by rursache** [comment]
9 years and counting...

**Comment by davidgodzsak** [comment]
When I do exactly this my code fails. I debugged it and found that the `methodLookup` always returns `PaymentMethod` (the parent class instead of the subclass type). How do I get around that? I guess this is more like a Swift problem, but I cannot find a solution to it.

Also 10th anniversary of this ticket this year 🎉

**Comment by davidgodzsak** [comment]
I think either I use Realm wrong or the [suggested workaround ](https://github.com/realm/realm-swift/issues/1109#issuecomment-1039160958) here is pretty terrible. So please point me to the right direction if I use this tool wrong. 

I use SwiftUI and a FlexibleSync config with `@AutoOpen`. With this solution I need to first of all subscribe all of the child types (in the example all the payment methods), otherwise I cannot add new ones to the database. This does not sound so bad at first, but it means that:
- I either need to mess up my data structures 
- Accept sub-optimal security and loads of unnecessary sync

Why?
I subscribe to the children classes in the realm, but cannot even specify [query conditions](https://www.mongodb.com/docs/realm/sdk/swift/sync/flexible-sync/#subscribe-to-queries) to what to sync, (e.g. only get payment methods that correspond to a specific purchase), because no data is stored in the child class regarding this.

So I can either mess up my data modelling by adding redundant data to the children that I use for filtering, or sync on all data in each child's table (which is unnecessary, bad idea in scenarios when data transmission is expensive, and could even have security concerns - loading stuff into the local db and memory that do not belong to the user).

Another flaw of this is that in the example we just create a new `Realm()` instance every time we want to resolve a value. With AutoOpen (and probable all other sane implementations) we need to pass in the `realm` instance to the `value` getter (of `AnyPaymentMethod` in this case).

If this is what is expected me to have polymorphism, then that's pretty bad and I cannot understand how this project could survive now almost 10 years with this ticket open.

**Comment by kpolleck** [comment]
Being just a little bit silly, should we plan a 10-year-old party for this request?

OK, OK...we are all in software development, and we know that there is a prioritization process for requirements--and it considers ROI, so difficult problems may not make the cut for a release.

I'm relatively new to using Realm and, in adding features to my application, I decided to refactor it to use subclasses that I want to store in Realm...thinking certainly that would be supported, no?  After all, it's supported in CoreData.  But, after a bit of refactoring and testing, I realized I should have researched that more in advance.

Thus, it seems to me that we need to make application architecture decisions based on what is available now (or maybe on a published roadmap), so I'm now rethinking my decision to use Realm at all!

**Comment by davidgodzsak** [comment]
welp, now they are deprecating Realm altogether :D

**Comment by rursache** [comment]
@jadar @davidgodzsak any link on Realm being deprecated?

**Comment by jadar** [comment]
@rursache https://www.mongodb.com/docs/atlas/device-sdks/deprecation/#std-label-device-sdks-deprecation

**Comment by bdkjones** [comment]
Realm Sync was the *ENTIRE* reason we chose Realm. Without the sync component, everyone is better served using Swift Data, the first-party solution that doesn't get suddenly terminated because some idiot MBA ran a spreadsheet.

**Comment by rursache** [comment]
SwiftData is a joke compared to Realm and even CoreData which actually wraps

**Comment by bdkjones** [comment]
@rursache You know what Core Data does NOT do? Randomly get nuked from orbit and cancelled. Call me crazy, but "continues to exist" is a pretty good feature for a product to have.

Realm is dead and irrelevant as of today. Zero reason to use it.

**Comment by marchy** [comment]
> @davidgodzsak Thanks for the heads up. To be fair, it sounds like the on-device SDKs will continue to exist. Just Realm Sync is deprecated and will be removed (which I never used, so that is a bit of complexity that can be deleted!) Maybe as open-source, we can get some better sync APIs that work with more technologies now!

Wow that is a pretty epic fail.

We chose to omit Realm Sync and built our entire sync layer by hand (both client-side and server-side) as to be able to have complete control over its many intricacies.

Bullet dodged on that one!

NOTE: We’ve since moved to GRDB over Realm as well due to the insane immutable-per-thread architecture Realm was built on, so wouldn’t have been an issue there either. But in both cases, the right answer ended up to go lower-level and not trust the black-box abstractions. At this point our ORM does the least possible thing - fetch records, not even relations between them, and we do all the object stitching ourselves. I wish it wasn’t this case, but reality just forced us otherwise.

**Comment by andrefmsilva** [comment]
@marchy 
> NOTE: We’ve since moved to GRDB over Realm as well due to the insane immutable-per-thread architecture Realm was built on, so wouldn’t have been an issue there either. But in both cases, the right answer ended up to go lower-level and not trust the black-box abstractions. At this point our ORM does the least possible thing - fetch records, not even relations between them, and we do all the object stitching ourselves. I wish it wasn’t this case, but reality just forced us otherwise.

Did you notice any performance or memory usage impact with GRDB?
